pkg/operator/ceph/cluster/osd/spec.go

```diff
@@ -61,6 +63,11 @@ const (
 	OSDOverPVCLabelKey                  = "ceph.rook.io/pvc"
 )
 
+var (
+	osdPodPrepareMemory = int64(display.MbTob(200))
```

Comment:
What are these numbers based on? If we aren't actually doing testing to see what resources are used, we should have some reason for our guess at what limit is acceptable.

I guessed that there might be 512 bytes of data structure for each disk in a node, and a node might have as many as 90 disks, which would be 45 kB. Having 2 MB per disk (4x my guess) does seem reasonable to me, but it would be nice to have something to suggest that 2 MB per disk is an appropriate guess.

Reply:
I don't think we should head the route of precise profiling, specifically because there might be cases where more memory is consumed and the pod can die (eg: encrypted scenario). So I've put 200 MB because it sounds reasonable to me.

Also, I don't think only basing your calculation the byte structure on the disk is enough, most of the disks do 4K nowadays, we also need to look at library opened and loaded in memory and subsequent forks.

I did a small test see: 

```
[root@minikube /]# /usr/bin/time -v ceph-volume lvm prepare --data /dev/vdd
Running command: /bin/ceph-authtool --gen-print-key
Running command: /bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring -i - osd new 4dcf2e97-cc34-4812-b2ca-b89cfa778ceb
Running command: /usr/sbin/vgcreate -s 1G --force --yes ceph-7b523d9a-8980-4273-ba75-6cbed9035028 /dev/vdd
 stderr:
 stderr: WARNING: Failed to connect to lvmetad. Falling back to device scanning.
 stderr:
 stdout: Physical volume "/dev/vdd" successfully created.
 stdout: Volume group "ceph-7b523d9a-8980-4273-ba75-6cbed9035028" successfully created
Running command: /usr/sbin/lvcreate --yes -l 100%FREE -n osd-block-4dcf2e97-cc34-4812-b2ca-b89cfa778ceb ceph-7b523d9a-8980-4273-ba75-6cbed9035028
 stderr:
 stderr: WARNING: Failed to connect to lvmetad. Falling back to device scanning.
 stderr:
 stdout: Logical volume "osd-block-4dcf2e97-cc34-4812-b2ca-b89cfa778ceb" created.
Running command: /bin/ceph-authtool --gen-print-key
Running command: /bin/mount -t tmpfs tmpfs /var/lib/ceph/osd/ceph-3
Running command: /usr/sbin/restorecon /var/lib/ceph/osd/ceph-3
Running command: /bin/chown -h ceph:ceph /dev/ceph-7b523d9a-8980-4273-ba75-6cbed9035028/osd-block-4dcf2e97-cc34-4812-b2ca-b89cfa778ceb
Running command: /bin/chown -R ceph:ceph /dev/mapper/ceph--7b523d9a--8980--4273--ba75--6cbed9035028-osd--block--4dcf2e97--cc34--4812--b2ca--b89cfa778ceb
Running command: /bin/ln -s /dev/ceph-7b523d9a-8980-4273-ba75-6cbed9035028/osd-block-4dcf2e97-cc34-4812-b2ca-b89cfa778ceb /var/lib/ceph/osd/ceph-3/block
Running command: /bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring mon getmap -o /var/lib/ceph/osd/ceph-3/activate.monmap
 stderr: got monmap epoch 3
 stderr:
Running command: /bin/ceph-authtool /var/lib/ceph/osd/ceph-3/keyring --create-keyring --name osd.3 --add-key AQCfnKVdo5FyDBAApH1s/68aV69QCwPGDj5kxw==
 stdout: creating /var/lib/ceph/osd/ceph-3/keyring
 stdout: added entity osd.3 auth(key=AQCfnKVdo5FyDBAApH1s/68aV69QCwPGDj5kxw==)
Running command: /bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-3/keyring
Running command: /bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-3/
Running command: /bin/ceph-osd --cluster ceph --osd-objectstore bluestore --mkfs -i 3 --monmap /var/lib/ceph/osd/ceph-3/activate.monmap --keyfile - --osd-data /var/lib/ceph/osd/ceph-3/ --osd-uuid 4dcf2e97-cc34-4812-b2ca-b89cfa778ceb --setuser ceph --setgroup ceph
--> ceph-volume lvm prepare successful for: /dev/vdd
        Command being timed: "ceph-volume lvm prepare --data /dev/vdd"
        User time (seconds): 4.84
        System time (seconds): 6.85
        Percent of CPU this job got: 80%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:14.46
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 41804
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 84721
        Voluntary context switches: 1713
        Involuntary context switches: 9122
        Swaps: 0
        File system inputs: 49528
        File system outputs: 5931
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

The interesting line is `Maximum resident set size (kbytes): 41804`, which is almost 42 Mb consumed by that process.
If you're really concern about the memory usage on the prepare pod I can probably lower to 100 Mb.

Thoughts?

Reply:
Something to consider is that we likely need different limits on the pods in two scenarios:
1. The per-node osd prepare jobs need to handle many devices in one job. 200MB does seem sufficient though.
1. For OSDs on PVCs we will create one prepare job per PVC. We could actually have a lot of prepare jobs that end up on the same node. Since there is only one OSD per job, the requirements could be lower. How about 100MB?

I think we should also consider setting different levels for the requests and the limits. We want to set lower requests so we make sure the pods get scheduled, but we can set higher limits so they don't get killed when they're running. They are only temporary jobs so it is fine if they burst higher. What if we set the limits to double the requests?

Reply:
We should have some way of overriding these hardcoded limits for cases where users might be seeing provisioning fail repeatedly due to k8s killing the pods for taking too much mem. Even an undocumented environment variable would be better than just hardcoding something.

Reply:
@travisn sounds good.
@BlaineEXE how do you suggest we should be doing this? Do we need to expose resources from the CRD for that job and then set value if nothing is specified?

Reply:
@leseb Yes, that approach occurred to me as well. What if we added a new category "prepareosd" to the [resources](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/cluster.yaml#L105-L116)? We could have a default and it would be overridden by the CR if set.